openai-triton-llvm: fix aarch64 and cross-compilation

Madouura opened the pull request:
## Description of changes
Fixes this comment: https://github.com/NixOS/nixpkgs/pull/263048#issuecomment-1807824236
Unsure how to test darwin, since it won't cross-compile.
This is why we're guarding ``openai-triton`` in ``torch``: 8bb0f732cf7638a0749a26fe37b60db1032daf55
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Madouura committed:
openai-triton-llvm: fix aarch64 and cross-compilation

Cross-compilation code taken/modified from LLVM 16 derivation

SomeoneSerge commented on pkgs/development/python-modules/torch/default.nix:
```diff
@@ -356,12 +356,11 @@ in buildPythonPackage rec {
     pybind11
   ]
+  # ROCm build and `torch.compile` requires openai-triton
+  ++ lib.optionals (!stdenv.isDarwin) [ openai-triton ]
```
What do you think about https://github.com/NixOS/nixpkgs/pull/263048#issuecomment-1806066394

Madouura replied:
I think we can do that.
I prefer ``enableTriton``, since it's possible it might be used for more than just ``torch.compile`` now or in the future.
I also think it would be best to enable it by default (or in this case, all but darwin) - There's no reason to exclude ``torch.compile`` by default.

Madouura replied:
Ended up going with ``tritonSupport`` to match the other ``...Support`` options.

SomeoneSerge commented on pkgs/by-name/op/openai-triton-llvm/package.nix:
```diff
@@ -69,57 +93,83 @@ stdenv.mkDerivation (finalAttrs: {
+    (lib.cmakeBool "LLVM_BUILD_TESTS" buildTests)
+  ] ++ lib.optionals (!isNative) (let
+    nativeToolchainFlags = let
+      nativeCC = pkgsBuildBuild.targetPackages.stdenv.cc;
```
Oh... could you briefly explain what is the difference between `pkgsBuildBuild.pkgsTargetTarget` and `pkgsBuildHost`? If I were trying to implement the cross support myself, how would I figure out I need to use this particular incantation?

Madouura replied:
If you meant ``pkgsBuildBuild`` and ``pkgsTargetTarget`` separately, then I believe ``pkgsBuildBuild`` is just the ``pkgs`` that the builder is using, and ``pkgsTargetTarget`` being the build target's ``pkgs``.
I'm not sure what the difference between ``pkgsTargetTarget`` and ``pkgsBuildHost`` would be though.
Or for that matter, with the two above, what the difference with ``pkgsHostHost`` would be.

Madouura replied:
Would ``pkgsBuildHost`` be the builder's target's ``pkgs``?
So confusing...

Madouura replied:
From what I can tell, ``pkgsTargetTarget`` is basically ``pkgsHostHost`` but for runtime or specialized stuff? Like a compiler?
I could be wrong, but you pretty much never want to use ``pkgsTargetTarget`` over ``pkgsHostHost``.
I don't fully understand it myself, but see: https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/cross-compilation.chapter.md

Madouura replied:
Specifically, ``pkgsTargetTarget`` seems to be used when the architecture can't be hot-plugged at runtime and you need to specify variations ahead of time in the derivation.
I think that should more or less resolve this question.

Madouura replied:
I somehow managed to not even look at the code this comment is on.
Okay, I think I understand why you asked this now.
The last comment still stands, and AFAIK this (``pkgsBuildBuild.targetPackages``) should be correct usage.
I wonder if it might be a legacy thing, but shouldn't ``pkgsBuildTarget`` work just as well?

Madouura committed:
python3Packages.torch: Add 'tritonSupport' option

Guard this option against 'stdenv.isDarwin', which currently is unknown whether it works or not.

Madouura commented:
cc @RaitoBezarius
Since you're on the LLVM team and this is using the cross-compiling solution from LLVM 16, if you have time I'd appreciate it if you looked over ``openai-triton-llvm`` and just made sure everything is happening like it should.
I haven't done much with cross-compilation.

RaitoBezarius commented:
Hi there, I can only review this in December as my bandwidth is severely low due to upcoming release and various engagements. I would recommend posting in the cross compilation channel.

Madouura commented:
The tests do pass on ``aarch64-linux``, so I won't worry too much about it. This can probably be merged.

> Hi there, I can only review this in December as my bandwidth is severely low due to upcoming release and various engagements. I would recommend posting in the cross compilation channel.

No problem, I'll post this PR there.

Artturin commented on pkgs/by-name/op/openai-triton-llvm/package.nix:
```diff
@@ -1,27 +1,49 @@
+  # but a later update will likely fix this (openai-triton-2.1.0)
+  python =
+    if buildTests
+    then python3Packages.python.withPackages (p: with p; [ psutil pygments pyyaml ])
```
```suggestion
    then python3Packages.python.pythonForBuild.withPackages (p: with p; [ psutil pygments pyyaml ])
```

No splicing with withPackages

Madouura commented on pkgs/by-name/op/openai-triton-llvm/package.nix:
```diff
@@ -69,57 +93,83 @@ stdenv.mkDerivation (finalAttrs: {
+    (lib.cmakeBool "LLVM_BUILD_TESTS" buildTests)
+  ] ++ lib.optionals (!isNative) (let
+    nativeToolchainFlags = let
+      nativeCC = pkgsBuildBuild.targetPackages.stdenv.cc;
```
Done.

Madouura commented:
I just realized I messed up badly with the cross code.
All of those should be "out", not "dev"/"lib".

Madouura commented:
Good thing I caught that, I guess tests only passed because they're not in ``installCheck``.

SomeoneSerge commented:
The diff LGTM, just running the builds now

```bash
❯ nix build github:Madouura/nixpkgs/pr/triton-llvm#pkgsCross.aarch64-multiplatform.python3Packages.torchWithRocm
error:
       … while evaluating the attribute 'drvPath'
...

         at /nix/store/zkjx4b3klwrsc3gk4pn3pi3j1hvqaws0-source/lib/customisation.nix:247:7:

...

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: Package ‘rocm-llvm-llvm-5.7.1’ in /nix/store/zkjx4b3klwrsc3gk4pn3pi3j1hvqaws0-source/pkgs/development/rocm-modules/5/llvm/base.nix:160 is marked as broken, refusing to evaluate.
```

Is this expected?

azuwis commented:
> @azuwis could you please test `nix build github:Madouura/nixpkgs/pr/triton-llvm#python3Packages.torch` on darwin?

`nix build github:Madouura/nixpkgs/pr/triton-llvm#python3Packages.torch` builds fine on aarch64-darwin.

Madouura commented:
> > @azuwis could you please test `nix build github:Madouura/nixpkgs/pr/triton-llvm#python3Packages.torch` on darwin?
> 
> `nix build github:Madouura/nixpkgs/pr/triton-llvm#python3Packages.torch` builds fine on aarch64-darwin.

Any chance you could test ``x86_64-darwin`` too?
If it works we can alter the ``torch`` commit.

azuwis commented:
> Any chance you could test `x86_64-darwin` too? If it works we can alter the `torch` commit. AFAIK the main blocking factor _should_ be `libpfm`, so if you could test with removing it too that'd be great.

I don't have a `x86_64-darwin` machine.

Madouura commented:
> > Any chance you could test `x86_64-darwin` too? If it works we can alter the `torch` commit. AFAIK the main blocking factor _should_ be `libpfm`, so if you could test with removing it too that'd be great.
> 
> I don't have a `x86_64-darwin` machine.

For now, let's keep darwin support guarded on torch then.
Once both are confirmed I think we can enable it.

azuwis commented:
@ofborg build python3Packages.torch

Madouura commented:
> @ofborg build python3Packages.torch

Why is torch failing on ``aarch64-darwin`` with ``openai-triton-llvm`` (???) being marked broken on it?
Any ``torch`` ``-darwin`` platforms shouldn't be pulling this. Is this an ofborg bug?

Artturin commented:
> > @ofborg build python3Packages.torch
> 
> Why is torch failing on `aarch64-darwin` with `openai-triton-llvm` (???) being marked broken on it? Any `torch` `-darwin` platforms shouldn't be pulling this. Is this an ofborg bug?

On this branch `python3Packages.torch` evals on darwin but on master it doesn't

SomeoneSerge commented:
Indeed, eval looks fine:

```
❯ nix eval github:Madouura/nixpkgs/pr/triton-llvm#legacyPackages.aarch64-darwin.python3Packages.torch.outPath
"/nix/store/rjil2kzg3jz0vqywbf8mlm7v4v2bmfw8-python3.11-torch-2.0.1"
❯ nix eval github:Madouura/nixpkgs/pr/triton-llvm#legacyPackages.x86_64-darwin.python3Packages.torch.outPath
"/nix/store/9afb5lvw0ifr9jcdn9ilv2hhlrwdil7p-python3.11-torch-2.0.1"
```


@Artturin why does ofborg care about whether torch evals on master?

Artturin commented:
> Indeed, eval looks fine:
> 
> ```
> ❯ nix eval github:Madouura/nixpkgs/pr/triton-llvm#legacyPackages.aarch64-darwin.python3Packages.torch.outPath
> "/nix/store/rjil2kzg3jz0vqywbf8mlm7v4v2bmfw8-python3.11-torch-2.0.1"
> ❯ nix eval github:Madouura/nixpkgs/pr/triton-llvm#legacyPackages.x86_64-darwin.python3Packages.torch.outPath
> "/nix/store/9afb5lvw0ifr9jcdn9ilv2hhlrwdil7p-python3.11-torch-2.0.1"
> ```
> 
> @Artturin why does ofborg care about whether torch evals on master?

It doesn't, I just said it because @Madouura thought it was broken on this branch too.

SomeoneSerge commented:
![image](https://github.com/NixOS/nixpkgs/assets/9720532/aeb218d8-1b0f-41e5-9b2c-61438ac56d5e)

Is this about master? Because I don't see how can it be about the current branch

Artturin commented:
@ofborg build python311Packages.torch

Madouura commented:
https://github.com/NixOS/nixpkgs/blob/332d8dc702c82b25f03523c7f62505befb3ea6d5/pkgs/development/python-modules/torch/default.nix#L292-L293
We can likely fix the ``aarch64-darwin`` error by adding the following under this:
```nix
"-Wno-error=maybe-uninitialized"
"-Wno-error=stringop-overflow"
```

I don't really want to over-complicate this PR with these kinds of fixes though and I have no way of actually testing darwin without annoying people with ofborg access until it works.

azuwis commented:
Did ofborg tried to merge master before build?

```
$ nix build github:Madouura/nixpkgs/pr/triton-llvm#python311Packages.torch
$ nix derivation show result/ | head                                                           
{
  "/nix/store/nqb2hfsn9vd06x3nhjgyqky0gsch99ck-python3.11-torch-2.0.1.drv": {
    "args": [
      "-e",
      "/nix/store/6xg259477c90a229xwmb53pdfkn6ig3g-default-builder.sh"
    ],
    "builder": "/nix/store/zzpm4317hn2y29rm46krsasaww9wxb1k-bash-5.2-p15/bin/bash",
    "env": {
      "BUILD_DOCS": "0",
      "BUILD_NAMEDTENSOR": "1",
```

But https://github.com/NixOS/nixpkgs/pull/267209/checks?check_run_id=18767769137 shows it tried to build /nix/store/wy6alr3i3883c43ir6n20w0x7112j0n9-python3.11-torch-2.0.1.drv

wegank commented:
From the error message on `aarch64-darwin`:

```
> [4235/4948] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/mkldnn/Pooling.cpp.o
> FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/mkldnn/Pooling.cpp.o
> ...
> error: unknown warning option '-Wno-maybe-uninitialized'; did you mean '-Wno-uninitialized'? [-Werror,-Wunknown-warning-option]
> error: unknown warning option '-Wno-stringop-overflow'; did you mean '-Wno-shift-overflow'? [-Werror,-Wunknown-warning-option]
> ninja: build stopped: subcommand failed.
```

`-Wno-maybe-uninitialized` and `-Wno-stringop-overflow` are not recognized by Clang. These lines having been added in #266995 without being tested, a fix would be to set them as `stdenv.cc.isGNU`-only.

(A commit named `python311Packages.torch: blabla` instead of `python3Packages.torch: blabla` would trigger ofborg builds to get the issue found, but anyway.)

The `x86_64-darwin` build times out at `[3871/5670]`, it doesn't even get to where the error occurs on `aarch64-darwin`. I have no idea if the build will succeed.

wegank committed:
python311Packages.pytorch: mark some flags as gcc-only
